### PR TITLE
many features and improvements

### PR DESCRIPTION
**Pull request by aajkl**
1. support ceph rbd
2. support ovn
3. support cloud-init
4. merge eru agent
5. many other improvements and bugfix

### CODE REVIEW

File: configs/codec.go [hunk 1]
```diff
@@ -5,13 +5,13 @@ import (
 
 	"github.com/BurntSushi/toml"
 
-	"github.com/projecteru2/yavirt/pkg/errors"
```
Comment (anrs) [inline configs/codec.go:8]: actually, why not wraps the `cockroachdb/errors` within `yavirt/pkg/errors`, then don't need to change anything from upper pkgs.

File: internal/eru/agent/agent.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+	labels := map[string]string{}
+	for _, label := range config.Labels {
+		parts := strings.Split(label, "=")
+		if len(parts) != 2 {
+			return nil, errors.Newf("invalid label %s", label)
+		}
+		labels[parts[0]] = parts[1]
+	}
+	go func() {
+		// Core need to connect to the local grpc server, so sleep 30s here to wait local grpc server up
```
Comment (anrs) [inline internal/eru/agent/agent.go:79]: sending a ping packet to be more graceful

File: internal/eru/agent/agent.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+			Endpoint: endpoint,
+			Podname:  config.Podname,
+			Labels:   labels,
+		}); err != nil {
+			e, ok := status.FromError(err)
+			if !ok {
+				logger.Error(ctx, err, "failed to add node")
+				return
+			}
+			if e.Code() == 1031 && strings.Contains(e.Message(), "node already exists") {
```
Comment (anrs) [inline internal/eru/agent/agent.go]: defines a var for (hardcoded) `1031` should be a good idea

File: internal/eru/agent/guest.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+			tcpChecker = append(tcpChecker, fmt.Sprintf("%s:%s", ip, port))
+		}
+	}
+	if healthCheck.HTTPPort != "" {
+		for _, ip := range g.IPs {
+			httpChecker = append(httpChecker, fmt.Sprintf("http://%s:%s%s", ip, healthCheck.HTTPPort, healthCheck.HTTPURL)) //nolint
+		}
+	}
+
+	f1 := utils.CheckHTTP(ctx, g.ID, httpChecker, healthCheck.HTTPCode, timeout)
```
Comment (anrs) [inline internal/eru/agent/guest.go]: how about shortcut here for saving cpu/bandwidth

File: internal/eru/agent/metrics.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func (e *MetricsCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- vmHealthyDesc
+}
+
+func (e *MetricsCollector) Collect(ch chan<- prometheus.Metric) {
+	for _, v := range e.wrkStatusCache.Items() {
+		wrkStatus, _ := v.Object.(*types.WorkloadStatus)
+		if !wrkStatus.Running {
```
Comment (anrs) [inline internal/eru/agent/metrics.go:52]: prefer to check `wrkStatus` if is nil in the first

File: internal/eru/agent/node.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+// HealthCheck.Timeout is used as timeout of requesting core Profile
+func (m *Manager) nodeStatusReport(ctx context.Context) {
+	logger := log.WithFunc("nodeStatusReport").WithField("hostname", m.config.Hostname)
+	logger.Debug(ctx, "report begins")
+	defer logger.Debug(ctx, "report ends")
+	if !m.svc.IsHealthy(ctx) {
+		logger.Warn(ctx, "service is not healthy")
+		return
+	}
+	if err := m.store.CheckHealth(ctx); err != nil {
```
Comment (anrs) [inline internal/eru/agent/node.go:44]: how about 
```
m.mCol.coreHealthy.Store(m.store.CheckHealth(ctx) == nil)
```
for removing the if/else

File: internal/eru/agent/workload.go [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	return status, nil
+}
+
+func (m *Manager) healthCheck(ctx context.Context) {
+	tick := time.NewTicker(time.Duration(m.config.HealthCheck.Interval) * time.Second)
+	defer tick.Stop()
+
+	_ = utils.Pool.Submit(func() { m.checkAllWorkloads(ctx) })
```
Comment (anrs) [inline internal/eru/agent/workload.go:117]: can remove `_ =` part from the statement

File: internal/eru/agent/workload.go [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+	for {
+		select {
+		case <-tick.C:
+			_ = utils.Pool.Submit(func() { m.checkAllWorkloads(ctx) })
```
Comment (anrs) [inline internal/eru/agent/workload.go:122]: also can remove the `_ =` part

File: internal/eru/agent/workload.go [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+	logger := log.WithFunc("checkAllWorkloads")
+	logger.Debug(ctx, "health check begin")
+	workloadIDs, err := m.ListWorkloadIDs(ctx)
+	if err != nil {
+		logger.Error(ctx, err, "error when list all workloads with label \"ERU=1\"")
+		return
+	}
+
+	for _, workloadID := range workloadIDs {
+		ID := workloadID
```
Comment (anrs) [inline internal/eru/agent/workload.go]: we can add comment to explain why we define a new var here.
(im not sure if is fixed the bug in golang higher version, if so, we can remove the new var)

File: internal/eru/agent/workload.go [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+		_ = utils.Pool.Submit(func() { m.checkOneWorkload(ctx, ID) })
```
Comment (anrs) [inline internal/eru/agent/workload.go]: dont need to `_ =` part

File: internal/eru/agent/workload.go [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+	logger := log.WithFunc("checkOneWorkload").WithField("ID", ID)
+	workloadStatus, err := m.GetStatus(ctx, ID, true)
+	if err != nil {
+		logger.Error(ctx, err, "failed to get status of workload")
+		return false
+	}
+
+	m.wrkStatusCache.Set(workloadStatus.ID, workloadStatus, 0)
+
+	if err = m.setWorkloadStatus(ctx, workloadStatus); err != nil {
```
Comment (anrs) [inline internal/eru/agent/workload.go]: prefer to a brand-new `err :=` here

File: internal/eru/recycle/recycle.go [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+	"github.com/projecteru2/yavirt/internal/eru/common"
+	"github.com/projecteru2/yavirt/internal/eru/store"
+	corestore "github.com/projecteru2/yavirt/internal/eru/store/core"
+	storemocks "github.com/projecteru2/yavirt/internal/eru/store/mocks"
+	"github.com/projecteru2/yavirt/internal/eru/types"
+	"github.com/projecteru2/yavirt/internal/service"
+	"github.com/projecteru2/yavirt/internal/utils"
+	"github.com/projecteru2/yavirt/internal/utils/notify/bison"
+	"github.com/samber/lo"
+	"google.golang.org/grpc/status"
```
Comment (anrs) [inline internal/eru/recycle/recycle.go:24]: prefer to group/separate pkgs from `3rd` and `yavirt`

File: internal/eru/recycle/recycle.go [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+		if err := svc.ControlGuest(ctx, utils.VirtID(eruID), virttypes.OpDestroy, true); err != nil {
+			logger.Errorf(ctx, err, "[recycle] failed to remove dangling guest %s", eruID)
+			return err
+		}
+		notifier := bison.GetService()
+		log.Debugf(ctx, "[recycle] notifier: %v", notifier)
+		if notifier != nil {
+			msgList := []string{
+				"<font color=#00CC33 size=10>delete dangling successfully </font>",
```
Comment (anrs) [inline internal/eru/recycle/recycle.go]: how about defining a template snippet here?

File: internal/eru/recycle/recycle.go [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(interval):
+		}
+
+		coreIDs, err := fetchWorkloads()
+		if err != nil {
+			continue
```
Comment (anrs) [inline internal/eru/recycle/recycle.go:121]: do we need to log the `err` but `not found` kind

File: internal/eru/recycle/recycle.go [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+		}
+		localIDs, err := svc.GetGuestIDList(context.Background())
+		if err != nil {
+			continue
+		}
+		coreMap := map[string]struct{}{}
+		for _, id := range coreIDs {
```
Comment (anrs) [inline internal/eru/recycle/recycle.go:128]: can we utilize the `lo.Map`

File: internal/eru/recycle/recycle.go [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+			eruID := virttypes.EruID(id)
+			if _, ok := coreMap[eruID]; ok {
+				continue
+			}
+			go deleteGuest(svc, eruID) //nolint
+		}
+	}
+}
+
+func Setup(ctx context.Context, cfg *configs.Config, t *testing.T) (err error) {
```
Comment (anrs) [inline internal/eru/recycle/recycle.go:141]: prefer to call this `Setup` and pass the store instance from unit test files

File: internal/eru/resources/core.go [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+	if err != nil {
+		logger.Errorf(ctx, err, "failed to fetch resource from core")
+		return
+	}
+	capacity := resp.Capacity
+	cm.mu.Lock()
+	defer cm.mu.Unlock()
+	for resName, rawParams := range capacity {
+		switch resName {
+		case "cpumem":
```
Comment (anrs) [inline internal/eru/resources/core.go]: strongly rec constants rather than literal in the switch/case

File: internal/eru/resources/core.go [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+func (cm *CoreResourcesManager) GetCpumem() (ans *cpumemtypes.NodeResource) {
+	cm.mu.Lock()
+	ans = cm.cpumem
+	cm.mu.Unlock()
+	if ans != nil {
+		return
+	}
+	cm.fetchResources()
```
Comment (anrs) [inline internal/eru/resources/core.go]: rename to `fetchResourcesWithLock` or `FetchResources` (with lock implicitly)
i'd like to the former and 
```
func (cm *CoreResourcesManager) GetCpumem() (ans *cpumemtypes.NodeResource) {
  cm.mu.Lock()
  defer cm.mu.Unlock()
  ...
}
```
only once

File: configs/codec.go (hunk 1 above)
Comment (aajkl) [reply configs/codec.go:8]: Using upstream package directly is enough and adding a thin wrapper is needless. actually I just want to unify the log and errors pkg of core,yavirt and some components developed by me.

File: internal/eru/agent/agent.go (hunk 2 above)
Comment (aajkl) [reply internal/eru/agent/agent.go:79]: ping what? local machine? sleeping 30s is enough here, it is more simple and works well

File: internal/eru/agent/agent.go (hunk 3 above)
Comment (aajkl) [reply internal/eru/agent/agent.go]: okay, I'll fix this

Comment (aajkl) [reply internal/eru/agent/agent.go]: these code numbers are from core, I just don't want to import too many dependencies, but I just read code and it's ok to import core here.

File: internal/eru/agent/node.go (hunk 6 above)
Comment (aajkl) [reply internal/eru/agent/node.go:44]: here I'll add a log message

File: internal/eru/agent/workload.go (hunk 7 above)
Comment (aajkl) [reply internal/eru/agent/workload.go:117]: removing `_=` will cause a lint issue

File: internal/eru/agent/workload.go (hunk 8 above)
Comment (aajkl) [reply internal/eru/agent/workload.go:122]: same as above

File: internal/eru/agent/workload.go (hunk 9 above)
Comment (aajkl) [reply internal/eru/agent/workload.go]: this is copied from agent, https://github.com/projecteru2/agent/blob/master/manager/workload/health_check.go#L41 I also don't know why defining a new var, I'll delete this

File: internal/eru/agent/workload.go (hunk 11 above)
Comment (aajkl) [reply internal/eru/agent/workload.go]: okay, this is also from agent

File: internal/eru/agent/metrics.go (hunk 5 above)
Comment (aajkl) [reply internal/eru/agent/metrics.go:52]: I'll add a check here, actually wrkStatus never be nil here

File: internal/eru/recycle/recycle.go (hunk 15 above)
Comment (aajkl) [reply internal/eru/recycle/recycle.go:128]: lo.Map returns slice and here is a map and i think for-loop is more intuitive here

File: internal/eru/agent/guest.go (hunk 4 above)
Comment (aajkl) [reply internal/eru/agent/guest.go]: this is specified by user and when checker list is empty, `checkHTTP` do nothing

File: internal/eru/resources/gpu.go [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+}
+
+func initGPUNameMap(gpuProdMapCfg map[string]string) error {
+	if len(gpuNameMap) > 0 {
+		return nil
+	}
+	if err := json.Unmarshal([]byte(gpuNameJSON), &gpuNameMap); err != nil {
+		return err
+	}
+	// merge name map from config to gpuNameMap
```
Comment (anrs) [inline internal/eru/resources/gpu.go:55]: `lo.Map` would help?

File: internal/eru/resources/gpu.go [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+			continue
+		}
+		totalCount += reqCount
+		singleTypeGPUs := g.gpuTypeMap[reqProd]
+		available := singleTypeGPUs.addrSet.Difference(usedAddrsMap)
+		if available.Cardinality() < reqCount {
+			return nil, errors.New("no enough GPU")
+		}
+		for addr := range available.Iter() {
+			if reqCount <= 0 {
```
Comment (anrs) [inline internal/eru/resources/gpu.go:107]:
```
if reqCount > 0 {
  for addr := range available.Iter() {
    ...
  }
}
```

File: internal/types/cloud_init.go [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+	Commands   []string          `json:"commands"`
+
+	MAC       string           `json:"-"`
+	CIDR      string           `json:"-"`
+	MTU       int              `json:"-"`
+	IFName    string           `json:"-"`
+	DefaultGW CloudInitGateway `json:"-"`
+}
+
+func initTpls() (err error) {
```
Comment (anrs) [inline internal/types/cloud_init.go:57]: how about `func init()` then we can remove all the `if xxx == nil` from the func

Comment (anrs) [reply internal/types/cloud_init.go:57]: or `sync.Once(initTpls)`

File: internal/eru/agent/workload.go (hunk 9 above)
Comment (anrs) [reply internal/eru/agent/workload.go]: can't be deleted because of https://go.dev/blog/loopvar-preview

Comment (aajkl) [reply internal/eru/agent/workload.go]: I see, just use `for idx := range workloadIDs` and define a new var in loop, so it's not that strange.

File: internal/types/cloud_init.go (hunk 21 above)
Comment (aajkl) [reply internal/types/cloud_init.go:57]: I don't like `init`,  I always run initialization code explicitly, `sync.Once` seems good

File: internal/eru/resources/gpu.go (hunk 19 above)
Comment (aajkl) [reply internal/eru/resources/gpu.go:55]: I think for-loop is very intuitive here, using functional style code here don't have any advantages.

**Comment by anrs** [comment]
@aajkl  can be merged after lint fixing

**Comment by aajkl** [comment]
@anrs done

**Pull request merged**